### ISSUE TITLE
Enhance un-aggregated timeserie storage

### ISSUE THREAD

**Issue by jd** [gnocchixyz/gnocchi#176]
The unaggregated (raw) time series that each metric has (the back window) is currently stored as one giant blob. Even if the back window is limited to e.g. 1 hour, this can still be millions of points down to the nanoseconds. This can make it huge, in theory, even if in practice it's never the case.

We should find a way to store it than a giant blob.

**Comment by chungg** [comment]
yeah, i noticed this... :(

i think this is a concern if backwindows are greater than a day.

**Comment by jd** [comment]
Here's a rough idea I've started to think about. I think this is similar to how a balanced tree works.

The idea is to split the time series into splits (again) with a maximum number of points _P_ per split.
A split is identified by its first timestamp (that's the _key_).

On insertion, the algorithm would do:
- List all splits and get them key (O(1))
- Retrieve the splits concerned by the new points (O(1))
- For each split:
  - Insert new points in the split
  - Check the size of the split:
    - If size(split) > _P_, then split the split in half and create 2 smaller splits.
  - Store back the split(s) (O(1))

That way each split size is between _P/2_ or _P_. What _P_ should be, I don't know.

On reading, listing all the splits with their keys would give the complete list of splits to retrieve based on their keys:
- List all splits and get them key (O(1))
- Retrieve splits needed for the time range asked (N*O(1))

All the operations are O(1). The main change compared to what is done currently is that:
- It adds a list() operation before insertion
- It can add a second write() operation on write
So from read()/write() we go to list()/read()/write()+(write())

For reading, we go from read() to list()+N*read()

Really a rough idea, might have issues, but I wanted to write it down. :)

**Comment by chungg** [comment]
is it O(1)? you'd need to list them and loop/search through to find key which is just smaller than the incoming timestamp. also, if it's a batch of measures where i assume you'd have to do the same for last incoming timestamp (or just take all splits after first?)

i think it'd be good if we define what the limits are of object size like we did for aggregate storage 3600*8B and also figure out the typical frequency of metrics are. obviously, it's not a point per 10 minute but also it's not every single nanosecond. maybe we're building a clever b-tree solution for an edge case.

**Comment by jd** [comment]
> is it O(1)? you'd need to list them and loop/search through to find key which is just smaller than the incoming timestamp.

Yeah searching in that list is probably O(log(n)) where n is the number of split. Though we can probably get some help from the backend. Though O(log(n)) for _this_ particular operation is not scary.

Let's say you pick _p_ = 7200 so even if you keep 1s granularity over 1y, that's 24*365=8760 splits in the worst case scenario where all splits are _p/2_ sized. That's O(log(n)) and in this case 14 local operations for searching. Should take a few milliseconds. :)

> also, if it's a batch of measures where i assume you'd have to do the same for last incoming timestamp (or just take all splits after first?)

Yeah you basically batch the operations, and retrieve all the concerned splits. It's the same, but faster since you can batch several operations at once.

> i think it'd be good if we define what the limits are of object size like we did for aggregate storage 3600*8B and also figure out the typical frequency of metrics are. obviously, it's not a point per 10 minute but also it's not every single nanosecond. maybe we're building a clever b-tree solution for an edge case.

Yes that's my bigger concern here. It's not clear what the size is, and it's also difficult to know the size in advance in the case that there's compression.

**Comment by chungg** [comment]
yeah, i agree, the 'n' we would be dealing with is probably going to be small. i imagine in most cases, as it's timeseries, most of the measures will be appending to the last key and you'll have odd case where you have a measure which comes late and it ends up creating a new key but the object will really only store one value

as devs, we know that this is a potential weak point in system as we understand the implementation. but i'm wondering if we know anyone that's hitting this. if we could somehow get info from real deployments? like a diagnostic tool, or more crudely, just trigger a warning log so we know "this will be a problem" rather than "this theoretically could be a problem"

**Comment by jd** [comment]
Yeah I don't think that'd be a real problem anyway. For an unaggregated timeseries, you can always go back and insert data. That's a problem for aggregated timeseries, because you need to know what to recompute. And if you don't want to go back too much in time, you have to have define a back window (rings a bell right).


I stumbled upon this paper:

http://www.cs.ucf.edu/~dcm/Teaching/COT4810-Spring2011/Literature/SplitOrderedLists.pdf

and it's quite interesting as it is a generalization of the idea I had here. I still need to double check what interesting we can steal from it, but sounds like the idea is workable.
